enterprise/reporting/flusher/flusher.go

```diff
@@ -0,0 +1,397 @@
+	}
+}
+
+func (f *Flusher) FlushAggressively(ctx context.Context) bool {
+	if !f.aggressiveFlushEnabled.Load() {
+		return false
+	}
+	lag := f.getLag(ctx)
+	reportingLagInMins := lag.Minutes()
+	return reportingLagInMins > f.lagThresholdForAggresiveFlushInMins.Load().Minutes()
+}
+
+// flush is the main logic for flushing data.
+func (f *Flusher) Flush(ctx context.Context) error {
+	currentUTC := time.Now().UTC()
+
+	// Emit the lag metric
+	lag := f.getLag(ctx)
+	f.flushLag.Gauge(lag.Seconds())
+
+	// 1. Get the time range to flush
+	s := time.Now()
+	start, end, valid, err := f.getRange(ctx, currentUTC)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return nil
+	}
+	f.minReportedAtQueryTimer.Since(s)
+
+	// 2. Aggregate reports. We have different aggregators for in-app and in-db aggregation
+	s = time.Now()
+	jsonReports, err := f.aggregate(ctx, start, end)
+	if err != nil {
+		return err
+	}
+	f.aggReportsTimer.Since(s)
+	f.aggReportsCounter.Observe(float64(len(jsonReports)))
+
+	// 3. Flush aggregated reports
+	s = time.Now()
+	err = f.send(ctx, jsonReports)
+	if err != nil {
+		return err
+	}
+	f.sendReportsTimer.Since(s)
+
+	// 4. Delete reports
+	s = time.Now()
+	if err := f.delete(ctx, start, end); err != nil {
+		return err
+	}
+	f.deleteReportsTimer.Since(s)
+
+	return nil
+}
+
+// Since we have hourly/daily/monthly aggregates on Reporting Service, we want the window to be within same hour
+// Don't consider most recent data where there are inserts happening
+// Always try to consider full window of flush interval or till current hour
+func (f *Flusher) getRange(ctx context.Context, currentUTC time.Time) (start, end time.Time, valid bool, err error) {
+	start, err = f.db.GetStart(ctx, f.table)
+	if err != nil {
+		return time.Time{}, time.Time{}, false, err
+	}
+
+	if start.IsZero() {
+		return start, end, false, nil
+	}
+
+	end = start.Add(f.flushWindow.Load())
+	nextHour := start.Truncate(time.Hour).Add(time.Hour)
+	endLimit := currentUTC.Add(-f.recentExclusionWindow.Load())
+
+	if end.After(nextHour) {
+		end = nextHour
+	}
+	if end.After(endLimit) {
+		end = endLimit
+	}
+
+	currentHourStart := currentUTC.Truncate(time.Hour)
+	if end.Sub(start) == f.flushWindow.Load() || end == currentHourStart {
+		return start, end, true, nil
+	}
+
+	return start, end, false, nil
+}
+
+func (f *Flusher) aggregate(ctx context.Context, start, end time.Time) ([]json.RawMessage, error) {
+	jsonReports, total, unique, err := f.aggregator.Aggregate(ctx, start, end)
+	if err != nil {
+		return nil, err
+	}
+	f.reportsCounter.Observe(float64(total))
+	f.aggReportsCounter.Observe(float64(unique))
+
+	return jsonReports, nil
+}
+
+func (f *Flusher) send(ctx context.Context, aggReports []json.RawMessage) error {
+	if f.batchSizeToReporting.Load() > 1 {
+		if err := f.sendInBatches(ctx, aggReports); err != nil {
+			return err
+		}
+	} else {
+		if err := f.sendIndividually(ctx, aggReports); err != nil {
+			return err
+		}
+	}
```

Comment:
```suggestion
		if err := f.sendInBatches(ctx, aggReports); err != nil {
			return err
		}
```

Can we treat `sendIndividually` as sendInBataches with batch size 1?